### PR TITLE
PrepareDedicatedHostGroup endpoint

### PR DESCRIPTION
**Pull request by SParekh**
1. Add functionality to invoke PrepareDedicatedHostGroup endpoint
This call to the Dedicated Host Manager library will prepare the host group by creating sufficient number of dedicated hosts so that a future call to VM or VMSS creation will be successful. 
2. Abstract Configurations to "config" class to handle complex Json configurations

### CODE REVIEW

File: DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs [hunk 1]
```diff
@@ -797,6 +784,162 @@ await Policy
...
+
+            await CreateDedicatedHosts();
+            
+            return dedicatedHosts ?? new List<DedicatedHost>();
+            
+            async Task<DedicatedHostGroup> GetDedicatedHostGroup()
+            {
+                try
+                {
+                    return (await computeManagementClient.DedicatedHostGroups.GetWithHttpMessagesAsync(resourceGroup, dhGroupName)).Body;
```
Comment (mdewan) [inline DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs:868]: Consider using Polly for retries here

File: DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs [hunk 2]
```diff
@@ -826,5 +969,92 @@ await Policy
...
+                foreach (var host in dhInFaultDomain)
+                {
+                    // Existing hosts can be different sku then DH Sku determined for VM size.
+                    availableVMCapacityInFaultDomain += (int)(host.InstanceView?.AvailableCapacity?.AllocatableVMs?
+                        .FirstOrDefault(v => v.VmSize.Equals(vmSku, StringComparison.InvariantCultureIgnoreCase))?.Count ?? 0);
+                }
+
+                if (vmsRequiredPerFaultDomain > availableVMCapacityInFaultDomain)
+                {
+                    dedicatedHosts.Add((fd, (int)(Math.Round(((decimal)vmsRequiredPerFaultDomain - availableVMCapacityInFaultDomain) / vmCapacityPerHost, MidpointRounding.ToPositiveInfinity))));
```
Comment (mdewan) [inline DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs:998]: Maybe add an intrumentation/logging message that new Hosts were allocated?

File: DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs [hunk 3]
```diff
@@ -826,5 +969,92 @@ await Policy
...
+                    2 => new int[] { 2 },
+                    _ => Enumerable.Range(0, dhGroupFaultDomainCount).ToArray() // As # of small, perf of using Range not significant
+                };
+            }
+        }
+
+        private (string dhSku, int vmCapacity) GetVmCapacityPerHost(string location, string vmSku)
+        {
+            var matchingConfig = _config.DedicatedHostConfigurationTable
+                .Where(c => (c.Location == "default" || c.Location.Equals(location, StringComparison.OrdinalIgnoreCase))
```
Comment (mdewan) [inline DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs:1024]: What's the purpose of the "default" check?

File: DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs (hunk 1 above)
Comment (SParekh) [reply DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs]: Implemented retry logic.

File: DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs (hunk 2 above)
Comment (SParekh) [reply DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs]: Information logged as suggested. Both at time of determining # of hosts required and also on successful creation of hosts.

File: DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs (hunk 3 above)
Comment (SParekh) [reply DedicatedHostsManager/DedicatedHostEngine/DedicatedHostEngine.cs:1021]: "DedicatedHostMapping" configuration defines mapping between host SKU, vm family, vm size. Some regions might have different capacity considerations and in such use case, region specific settings would be provided. If none provided, "default" settings would be applied. Requirement specified this as sample
vm_family_to_host_sku= {"DSv3":{"Default":"DSv3-Type1", "ES East":"DSv3-Type2"},
                        "ESv3":{"Default":"ESv3-Type1", "ES East":"ESv3-Type2"},…} 

Currently ONLY default defined in configuration sample provided for "DSv3-Type1" as it is standard across all regions.

**Pull request merged**